### PR TITLE
feat: Add additional metrics for shuffle write

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

I would like to understand how much time is spent on shuffle writing.

## What changes are included in this PR?

- Introduce a new `write_time` native metric to record write time instead of using `elapsed_time`
- Use `elapsed_time` to measure total native time (excluding executing the child plan and fetching data)
- Add new `input_time` native metric for measuring the time for `ShuffleWriterExec` to execute the child plan and fetch its input data
- Add new `shuffleWallTime` JVM metric to measure total time of shuffle
- Update docs

## Spark UI

Note the new metrics:

- "native shuffle time"
- "native shuffle input time"
- "shuffle wall time"

Screenshot from 2024-12-16 15-21-24

## Native plan

```
ShuffleWriterExec: ..., metrics=[elapsed_compute=42.425493ms, ..., input_time=255.873165ms, write_time=462.517µs]
  ScanExec: source=[ShuffleWriterInput], metrics=[elapsed_compute=254.994282ms, ...]
```

### CODE REVIEW

File: docs/source/user-guide/metrics.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+# Comet Metrics
+
+## Spark SQL Metrics
+
+Set `spark.comet.metrics.detailed=true` to see all available Comet metrics.
+
+### CometScanExec
+
+`CometScanExec` uses nanoseconds for total scan time. Spark also measures scan time in nanoseconds but converts to
```
Comment (kazuyukitanimura) [inline docs/source/user-guide/metrics.md:28]: This sounds like a problem statement. Did you mean that `spark.comet.metrics.detailed=true` will not loose the precision?

**Review by kazuyukitanimura — APPROVED** [review]

**Comment by andygrove** [comment]
Also added `ipc_time`.

Screenshot from 2024-12-16 16-58-39

File: docs/source/user-guide/metrics.md (hunk 1 above)
Comment (andygrove) [reply docs/source/user-guide/metrics.md]: Thanks, I have updated this

Comment (andygrove) [reply docs/source/user-guide/metrics.md]: Spark converts nanos to millis on each batch:

```scala
        override def hasNext: Boolean = {
          // The `FileScanRDD` returns an iterator which scans the file during the `hasNext` call.
          val startNs = System.nanoTime()
          val res = batches.hasNext
          scanTime += NANOSECONDS.toMillis(System.nanoTime() - startNs)
          res
        }
```

We just use the nano time:

```
        override def hasNext: Boolean = {
          // The `FileScanRDD` returns an iterator which scans the file during the `hasNext` call.
          val startNs = System.nanoTime()
          val res = batches.hasNext
          scanTime += System.nanoTime() - startNs
          res
        }
```

It actually makes a large difference to the time reported in some cases

**Comment by andygrove** [comment]
Seeing some segmentation faults in CI:

```
# Problematic frame:
# C  [libcomet-9228012970092653184.so+0x274908f]  core::sync::atomic::AtomicUsize::fetch_add::h50346ebf04f5a2ef+0x4f
```

**Comment by andygrove** [comment]
I may have found a bug in DataFusion

```
 # A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fb91a124f46, pid=762, tid=833
#
# JRE version: OpenJDK Runtime Environment Zulu11.76+21-CA (11.0.25+9) (build 11.0.25+9-LTS)
# Java VM: OpenJDK 64-Bit Server VM Zulu11.76+21-CA (11.0.25+9-LTS, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [libcomet-1968252056067325991.so+0x1524f46]  datafusion_physical_plan::metrics::value::ScopedTimerGuard::stop::ha579d5e7b5f1a919+0x46
```

File: docs/source/user-guide/metrics.md (hunk 1 above)
Comment (andygrove) [reply docs/source/user-guide/metrics.md]: I have updated the description in the metrics guide to explain this in more detail.

**Comment by andygrove** [comment]
I created a new simpler PR to replace this one: https://github.com/apache/datafusion-comet/pull/1175

File: docs/source/user-guide/metrics.md (hunk 1 above)
Comment (parthchandra) [reply docs/source/user-guide/metrics.md]: I see. That could be a big difference in small datasets (wonder if the same occurs when we have large files). Either way, better to not lose precision. We are not likely to run into overflow issues are we?

**Pull request closed without merging**